app/src/forms/form/service.js

```diff
@@ -0,0 +1,161 @@
+
+      await Form.query(trx).insert(obj);
+      await trx.commit();
+      const result = await service.readForm(obj.id);
+      return result;
+    } catch (err) {
+      if (trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  updateForm: async (formId, data) => {
+    let trx;
+    try {
+      const obj = await service.readForm(formId);
+      trx = await transaction.start(Form.knex());
+
+      await Form.query(trx).patchAndFetchById(formId, {name: data.name, description: data.description, public: data.public, active: data.active, labels: data.labels});
+      await trx.commit();
+      const result = await service.readForm(obj.id);
+      return result;
+    } catch (err) {
+      if (trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  readForm: async (formId) => {
+    return Form.query()
+      .findById(formId)
+      .allowGraph('[versions]')
+      .withGraphFetched('versions(orderVersionDescending)')
+      .throwIfNotFound();
+  },
+
+  listVersions: async (formId) => {
+    return FormVersion.query()
+      .where('formId', formId)
+      .modify('orderVersionDescending');
+  },
+
+  createVersion: async (formId, data) => {
+    let trx;
+    try {
+      const form = await service.readForm(formId);
+      trx = await transaction.start(FormVersion.knex());
+
+      const obj = Object.assign({}, data);
+      obj.id =  uuidv4();
+      obj.formId = form.id;
+      obj.version = form.versions.length ? form.versions[0].version + 1 : 1;
+
+      await FormVersion.query(trx).insert(obj);
+      await trx.commit();
+      const result = await service.readVersion(obj.id);
+      return result;
+    } catch (err) {
+      if (trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  updateVersion: async (formVersionId, data) => {
+    let trx;
+    try {
+      const obj = await service.readVersion(formVersionId);
+      trx = await transaction.start(FormVersion.knex());
+
+      // check if in draft mode?
+
+      await FormVersion.query(trx).patchAndFetchById(formVersionId, {draft: data.draft, schema: data.schema});
+      await trx.commit();
+      const result = await service.readVersion(obj.id);
+      return result;
+    } catch (err) {
+      if (trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  readVersion: async (formVersionId) => {
+    return FormVersion.query()
+      .findById(formVersionId)
+      .throwIfNotFound();
+  },
+
+  listSubmissions: async (formVersionId) => {
+    return FormSubmission.query()
+      .where('formVersionId', formVersionId)
+      .modify('orderDescending');
+  },
+
+  createSubmission: async (formVersionId, data) => {
+    let trx;
+    try {
+      const formVersion = await service.readVersion(formVersionId);
+      trx = await transaction.start(FormSubmission.knex());
+
+      const obj = Object.assign({}, data);
+      obj.id = uuidv4();
+      obj.formVersionId = formVersion.id;
+      obj.confirmationId = obj.id.substring(0, 8).toUpperCase();
```

Comment:
Wondering if we should still be retaining a submission/confirmation id of just 8 characters or if we should be using the whole GUID instead. The confirmationid shorthand is slightly less useful if we don't have a way to reverse lookup and find your original submission otherwise.

Reply:
yes, confirmation id is still valuable.  rendered on screen, it's easier to digest, plus we format it and save it so the the UI doesn't have to do any mangling.